### PR TITLE
Make the STJ source generator cache friendly

### PR DESCRIPTION
**Pull request by eiriktsarpalis**
This PR attempts to improve the cache friendliness of the incremental source generator by making the following changes:

1. Removes all `SourceProductionContext` dependencies from the `Parser` class.
2. Removes all `Compilation` and `ISymbol` dependencies from the `Emitter` class.
3. Makes the intermediate `SourceGenerationSpec` model type immutable and structurally equatable so that it can be cached incrementally.

Contributes to https://github.com/dotnet/runtime/issues/68353.

### CODE REVIEW

File: src/libraries/System.Text.Json/gen/ContextGenerationSpec.cs [hunk 1]
```diff
@@ -1,46 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
-// The .NET Foundation licenses this file to you under the MIT license.
-
-using System.Collections.Generic;
-using System.Text.Json.Serialization;
-using System.Text.Json.Reflection;
-using System.Diagnostics;
-using Microsoft.CodeAnalysis;
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/ContextGenerationSpec.cs:8]: This file has been moved under the `Model` folder, but git seems to think it's been deleted.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 2]
```diff
@@ -76,29 +76,23 @@ private sealed partial class Emitter
             private const string JsonTypeInfoTypeRef = "global::System.Text.Json.Serialization.Metadata.JsonTypeInfo";
             private const string JsonTypeInfoResolverTypeRef = "global::System.Text.Json.Serialization.Metadata.IJsonTypeInfoResolver";
 
-            private static DiagnosticDescriptor TypeNotSupported { get; } = new DiagnosticDescriptor(
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:79]: Descriptors moved to the `Parser` class.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 3]
```diff
@@ -1482,7 +1392,176 @@ string GetParamDefaultValueAsString(object? value, Type type, string typeRef)
                         return FormatNumber();
                 }
 
-                string FormatNumber() => $"({typeRef})({Convert.ToString(value, CultureInfo.InvariantCulture)})";
+                string FormatNumber() => $"({type.FullyQualifiedName})({Convert.ToString(value, CultureInfo.InvariantCulture)})";
+            }
+
+            private static bool ShouldGenerateMetadata(TypeGenerationSpec typeSpec)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1398]: Methods from here on have been moved over from the `TypeGenerationSpec` class.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 4]
```diff
@@ -112,26 +111,41 @@ private sealed class Parser
             // Needed for converter validation
             private readonly Type? _jsonConverterOfTType;
 
+            // Keeps track of generated context type names
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:114]: Moved from the Emitter class.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 5]
```diff
@@ -1142,36 +1121,55 @@ void CacheMemberHelper(Location memberLocation)
...
-                    canContainNullableReferenceAnnotations: canContainNullableReferenceAnnotations,
-                    hasTypeFactoryConverter: hasTypeFactoryConverter,
-                    hasPropertyFactoryConverters: hasPropertyFactoryConverters,
-                    isPolymorphic: isPolymorphic);
-
-                return typeMetadata;
+                var typeRef = new TypeRef(type);
+                typeInfoPropertyName ??= type.GetTypeInfoPropertyName();
+
+                if (classType is ClassType.TypeUnsupportedBySourceGen)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1127]: Diagnostic reports moved over from the Emitter class.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 6]
```diff
@@ -1266,6 +1265,11 @@ bool MatchesConstructorParameter(ParameterGenerationSpec paramSpec)
                     out bool setterIsVirtual,
                     out bool setterIsInitOnly);
 
+                if (!isPublic && !memberType.IsPublic)
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1268]: Checks extracted from the callsite.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs [hunk 7]
```diff
@@ -787,31 +637,198 @@ public class ClassWithPropertiesAndFieldsThatAreReservedKeywords
...
+            ImmutableArray<Diagnostic> generatorDiags = result.NewCompilation.GetDiagnostics();
 
             // No diagnostics expected.
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Info, generatorDiags, Array.Empty<(Location, string)>());
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Warning, generatorDiags, Array.Empty<(Location, string)>());
             CompilationHelper.CheckDiagnosticMessages(DiagnosticSeverity.Error, generatorDiags, Array.Empty<(Location, string)>());
         }
+
+        [Theory]
+        [MemberData(nameof(GetCompilationHelperFactories))]
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:668]: New tests validating cacheability properties of the model types.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/TypeWrapperTests.cs [hunk 8]
```diff
@@ -44,256 +43,88 @@ public class ReferencedType
...
-            Assert.Empty(newCompilation.GetDiagnostics().Where(diag => diag.Severity.Equals(DiagnosticSeverity.Error)));
+            Assert.Empty(result.Diagnostics.Where(diag => diag.Severity.Equals(DiagnosticSeverity.Error)));
+            Assert.Empty(result.NewCompilation.GetDiagnostics().Where(diag => diag.Severity.Equals(DiagnosticSeverity.Error)));
 
             // Should find both types since compilation above was successful.
-            Assert.Equal(2, generator.GetSerializableTypes().Count);
-        }
-
-        [Fact]
-        public void CanGetAttributes()
```
Comment (eiriktsarpalis) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/TypeWrapperTests.cs:84]: Test deleted since the source generator no longer surfaces "TypeWrapper" values. The reflection wrappers will eventually get removed in a follow-up PR.

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs (hunk 7 above)
Comment (jkoritzinsky) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:668]: There's also Roslyn APIs that explicitly validate that a given run of the generator is incremental with respect to the previous run. I think it's worthwhile to add tests using that API as well (and you can build these tests to run on that API instead of building something from scratch if you would like).

Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:668]: Interesting, could you point me to examples?

Comment (jkoritzinsky) [reply src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/JsonSourceGeneratorTests.cs:668]: Here's the examples for the LibraryImportGenerator: https://github.com/dotnet/runtime/blob/main/src/libraries/System.Runtime.InteropServices/tests/LibraryImportGenerator.UnitTests/IncrementalGenerationTests.cs

Here's some examples in Roslyn: https://github.com/dotnet/roslyn/blob/236149d501afb0e898069243a6418435e9140ee0/src/Compilers/CSharp/Test/Semantic/SourceGeneration/GeneratorDriverTests.cs#L1665

Basically you can name a step and then traverse the tree of steps to see the inputs to a step and its outputs and whether or not they were new/modified/unchanged/pulled from the cache/removed/etc.

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+using System.Collections;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
+    /// Provides an immutable list implementation which implements sequence equality.
+    /// </summary>
+    public sealed class ImmutableEquatableArray<T> : IEquatable<ImmutableEquatableArray<T>>, IReadOnlyList<T>
```
Comment (Tornhoof) [inline src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:14]: you might want to add a constraint like where T : IEquatable<T>, because the whole caching idea only works if you have proper structural equality and not the normal ref equality, as you wrote in the comment in the other file. I saw that you converted all those Specs to records, which auto-implement IEquality anyway.

File: src/libraries/System.Text.Json/gen/Model/ParameterGenerationSpec.cs [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+    ///
+    /// When adding new members to the type, please ensure that these properties
+    /// are satisfied otherwise we risk breaking incremental caching in the source generator!
+    /// </remarks>
+    public sealed record ParameterGenerationSpec
+    {
+        public required TypeRef ParameterType { get; init; }
+        public required string Name { get; init; }
+        public required bool HasDefaultValue { get; init; }
+        public required object? DefaultValue { get; init; }
```
Comment (Tornhoof) [inline src/libraries/System.Text.Json/gen/Model/ParameterGenerationSpec.cs:28]: You might want to add a comment to object? DefaultValue, why your nice comment above does not apply to it, i.e. default Values need to be constant expressions or value types and such fulfil your requirement

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs [hunk 11]
```diff
@@ -0,0 +1,74 @@
...
+
+        public override bool Equals(object? obj)
+            => obj is ImmutableEquatableArray<T> other && Equals(other);
+
+        public override int GetHashCode()
+        {
+            int hash = 0;
+            foreach (T value in _values)
+            {
+                Combine(hash, value is null ? 0 : value.GetHashCode());
```
Comment (Tornhoof) [inline src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:36]: you're never assigning hash again.

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs (hunk 9 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:14]: Good idea, thanks.

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs (hunk 11 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs]: Good catch, thanks!

File: src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs [hunk 12]
```diff
@@ -10,12 +10,7 @@ namespace System.Text.Json.Serialization
     /// a property or field is ignored during serialization and deserialization. This option
     /// overrides the setting on <see cref="JsonSerializerOptions.DefaultIgnoreCondition"/>.
     /// </summary>
-#if BUILDING_SOURCE_GENERATOR
-    internal
-#else
-    public
-#endif
-    enum JsonIgnoreCondition
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:18]: Why is this changing?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 13]
```diff
@@ -1482,7 +1392,176 @@ string GetParamDefaultValueAsString(object? value, Type type, string typeRef)
...
+                => IsGenerationModeSpecified(typeSpec, JsonSourceGenerationMode.Metadata);
+
+            private static bool ShouldGenerateSerializationLogic(TypeGenerationSpec typeSpec)
+                => IsGenerationModeSpecified(typeSpec, JsonSourceGenerationMode.Serialization) && IsFastPathSupported(typeSpec);
+
+            private static bool IsGenerationModeSpecified(TypeGenerationSpec typeSpec, JsonSourceGenerationMode mode)
+                => typeSpec.GenerationMode == JsonSourceGenerationMode.Default || (mode & typeSpec.GenerationMode) != 0;
+
+
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1407]:
```suggestion
```

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 14]
```diff
@@ -136,22 +138,24 @@ public void Emit()
                         isRootContextDef: true);
 
                     // Add GetJsonTypeInfo override implementation.
-                    AddSource($"{contextName}.GetJsonTypeInfo.g.cs", GetGetTypeInfoImplementation(), interfaceImplementation: JsonTypeInfoResolverTypeRef);
+                    AddSource($"{contextName}.GetJsonTypeInfo.g.cs", GetGetTypeInfoImplementation(contextGenerationSpec), interfaceImplementation: JsonTypeInfoResolverTypeRef);
 
-                    // Add property name initialization.
+                    // Add property name initialization. TODO don't generate if empty set.
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:143]: I plan on following up with more changes to the individual classes after this decoupling PR is merged. I was going to remove the comment but I forgot :-)

File: src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs (hunk 12 above)
Comment (stephentoub) [reply src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:18]: If it's necessary for unit testing could we leave them as internal and use InternalsVisibleTo from the unit tests?  It's probably not a huge deal, but exposing public surface area from the source generator does create an opportunity for someone to reference it and depend on these.  We intend to not support that, but... yeah.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs [hunk 15]
```diff
@@ -30,42 +31,35 @@ public void Initialize(IncrementalGeneratorInitializationContext context)
...
                     (context, _) => (ClassDeclarationSyntax)context.TargetNode);
 
-            IncrementalValueProvider<(Compilation, ImmutableArray<ClassDeclarationSyntax>)> compilationAndClasses =
-                context.CompilationProvider.Combine(classDeclarations.Collect());
+            IncrementalValueProvider<SourceGenerationSpec?> sourceGenSpec = context.CompilationProvider
+                .Combine(classDeclarations.Collect())
+                .Select(static (tuple, cancellationToken) =>
+                {
+                    Parser parser = new(tuple.Left);
+                    return parser.GetGenerationSpec(tuple.Right, cancellationToken);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:39]: Nit: both for the parser and the emitter, would it be cleaner to just have an exposed static method?  Seems like the use is effectively `new Parser(...).GetGenerationSpec(...)`, with the caller not actually needing the instance reference other than for invoking the single method.

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs [hunk 16]
```diff
@@ -0,0 +1,74 @@
...
+            private readonly T[] _values;
+            private int _index;
+
+            internal Enumerator(T[] values)
+            {
+                _values = values;
+                _index = -1;
+            }
+
+            public bool MoveNext() => ++_index < _values.Length;
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:58]: We wouldn't do this if this were a publicly consumable type (call MoveNext enough after it returns false and it could start returning true again).  But presumably even though this is public the intent here is that no one is using it other than us?

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs [hunk 17]
```diff
@@ -0,0 +1,74 @@
...
+            {
+                hash = HashHelpers.Combine(hash, value is null ? 0 : value.GetHashCode());
+            }
+
+            return hash;
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(_values);
+        IEnumerator<T> IEnumerable<T>.GetEnumerator() => ((IEnumerable<T>)_values).GetEnumerator();
+        IEnumerator IEnumerable.GetEnumerator() => ((IEnumerable<T>)_values).GetEnumerator();
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:45]: Does this need the cast?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs (hunk 15 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Roslyn4.0.cs:39]: Yes, I plan to do this on follow-up PRs (and also remove all the System.Reflection wrappers over the Symbols API) after this one is merged.

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs [hunk 18]
```diff
@@ -0,0 +1,74 @@
...
+    public static class ImmutableEquatableArray
+    {
+        public static ImmutableEquatableArray<T> Empty<T>() where T : IEquatable<T>
+            => ImmutableEquatableArray<T>.Empty;
+
+        public static ImmutableEquatableArray<T> ToImmutableEquatableArray<T>(this IEnumerable<T> values) where T : IEquatable<T>
+            => new(values);
+
+        public static ImmutableEquatableArray<T> Create<T>(params T[] values) where T : IEquatable<T>
+            => values is null or { Length: 0 } ? ImmutableEquatableArray<T>.Empty : new(values);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:72]:
```suggestion
            => values is { Length: > 0 } ? new(values) : ImmutableEquatableArray<T>.Empty;
```
would let you avoid the extra expressed null check

File: src/libraries/System.Text.Json/gen/Model/TypeRef.cs [hunk 19]
```diff
@@ -0,0 +1,51 @@
...
+        /// </summary>
+        public string FullyQualifiedName { get; }
+
+        public bool IsValueType { get; }
+        public TypeKind TypeKind { get; }
+        public SpecialType SpecialType { get; }
+
+        // The spec is derived from cached `System.Type` instances, which are generally annotation-agnostic.
+        // Hence we can only record the potential for nullable annotations being possible for the runtime type.
+        // TODO: consider deriving the generation spec from the Roslyn symbols directly.
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Model/TypeRef.cs:42]: Another TODO comment

File: src/libraries/System.Text.Json/gen/Reflection/RoslynExtensions.cs [hunk 20]
```diff
@@ -33,6 +33,13 @@ public static IEnumerable<INamedTypeSymbol> BaseTypes(this INamedTypeSymbol type
             }
         }
 
+        /// <summary>
+        /// Creates a copy of the Location instance that does not capture a reference to Compilation.
+        /// </summary>
+        [return: NotNullIfNotNull(nameof(location))]
+        public static Location? GetTrimmedLocation(this Location? location)
+            => location is null ? null : Location.Create(location.SourceTree?.FilePath ?? "", location.SourceSpan, location.GetLineSpan().Span);
```
Comment (stephentoub) [inline src/libraries/System.Text.Json/gen/Reflection/RoslynExtensions.cs:41]: We've now duplicated this in several of our generators. @chsienki, should Roslyn include helpers for this (and maybe more generally on a Diagnostic so that someone writing a source generator can easily avoid referencing the compilation from a Diagnostic they hold onto)?

**Review by stephentoub — APPROVED** [review]
Have you validated this has the desired impact on recomputation in VS?

File: src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs (hunk 12 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:18]: I discussed this with @ericstj and @layomia and we figured that these wouldn't be public in practice since there's no easy way for users to reference them directly. The source generator classes themselves are public already but we don't particularly care about keeping them backward compatible. The SDK assemblies appear to also have public types for the same reason. I'd rather we fall back to IVT if there's strong pushback to the idea of making the model types public.

cc @ViktorHofer

File: src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs (hunk 16 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Model/ImmutableEquatableArray.cs:58]: I could try improving this a bit, since most likely we'd need to reuse that implementation for other source generators as well.

**Comment by eiriktsarpalis** [comment]
> Have you validated this has the desired impact on recomputation in VS?

Yes, both in unit testing and debugging VS itself. Curiously enough though it appears to be re-triggering periodically every ~20 seconds or so regardless of whether it is needed or not. @chsienki @sharwell is this expected behavior?

File: src/libraries/System.Text.Json/gen/Model/ClassType.cs [hunk 21]
```diff
@@ -1,13 +1,9 @@
...
 // The .NET Foundation licenses this file to you under the MIT license.
 
-using System;
-using System.Collections.Generic;
-using System.Text;
-
 namespace System.Text.Json.SourceGeneration
 {
-    internal enum ClassType
+    public enum ClassType
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/Model/ClassType.cs:6]: nit: we can remove the ctors part of line 9. Also this is tangential but at this point are there any types that we don't emit source for? I remember multi-dim arrays but we might have switched to generating metadata that throws `NotSupportedException` in some way.

File: src/libraries/System.Text.Json/gen/Model/ParameterGenerationSpec.cs [hunk 22]
```diff
@@ -0,0 +1,34 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+namespace System.Text.Json.SourceGeneration
+{
+    /// <summary>
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/Model/ParameterGenerationSpec.cs:6]: Could <inheritdoc/> be used here to avoid duplication and future divergence?

File: src/libraries/System.Text.Json/gen/Model/TypeRef.cs [hunk 23]
```diff
@@ -0,0 +1,50 @@
...
+        {
+            Debug.Assert(type is TypeWrapper);
+            var wrapper = ((TypeWrapper)type);
+
+            Name = type.Name;
+            FullyQualifiedName = type.GetCompilableName();
+            IsValueType = wrapper.IsValueType;
+            TypeKind = wrapper.Symbol.TypeKind;
+            SpecialType = wrapper.Symbol.SpecialType;
+            CanContainNullableReferenceAnnotations = type.CanContainNullableReferenceTypeAnnotations();
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/Model/TypeRef.cs:26]: Why not a record for this type? Seems like pointwise equality still holds given these members.

File: src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs (hunk 12 above)
Comment (ViktorHofer) [reply src/libraries/System.Text.Json/Common/JsonIgnoreCondition.cs:18]: I don't have a strong preference here just wanted to share that I noticed some issues with it. I.e. winforms heavily depends on IVT for their testing and when adding polyfills to netstandard2.0 like NullableAttributes.cs ([which runtime does implicitly](https://github.com/dotnet/runtime/blob/93a811f1ef273e64bd2a2280bce50dfdd8be6fc9/src/libraries/Directory.Build.targets#L165-L171) you get compiler errors. That's because the internal types are visible by the consumer which clash with the ones provided via the framework, i.e. when targeting .NETCoreApp.

Unsure if our source generator tests reference the source generator output assemblies here though.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs [hunk 24]
```diff
@@ -382,18 +389,8 @@ public Parser(Compilation compilation, in JsonSourceGenerationContext sourceGene
...
-                    CharType = _charType,
-                    DateTimeType = _dateTimeType,
-                    DateTimeOffsetType = _dateTimeOffsetType,
-                    GuidType = _guidType,
-                    StringType = _stringType,
-                    NumberTypes = _numberTypes,
-#if DEBUG
-                    MetadataLoadContext = _metadataLoadContext,
-#endif
+                    ContextGenerationSpecs = contextGenSpecList.ToImmutableEquatableArray(),
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:392]: nit: match the names
```suggestion
                    ContextGenerationSpecs = contextGenSpecs.ToImmutableEquatableArray(),
```

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 6 above)
Comment (layomia) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:1268]: Is this a bug fix?

File: src/libraries/System.Text.Json/gen/Reflection/RoslynExtensions.cs (hunk 20 above)
Comment (chsienki) [reply src/libraries/System.Text.Json/gen/Reflection/RoslynExtensions.cs:41]: Yeah, that seems like a good candidate for a new API

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 25]
```diff
@@ -1482,7 +1392,175 @@ string GetParamDefaultValueAsString(object? value, Type type, string typeRef)
                         return FormatNumber();
                 }
 
-                string FormatNumber() => $"({typeRef})({Convert.ToString(value, CultureInfo.InvariantCulture)})";
+                string FormatNumber() => $"({type.FullyQualifiedName})({Convert.ToString(value, CultureInfo.InvariantCulture)})";
+            }
+
+            private static bool ShouldGenerateMetadata(TypeGenerationSpec typeSpec)
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1398]: Should the bool value be a property on `TypeGenerationSpec`?

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 26]
```diff
@@ -1482,7 +1392,175 @@ string GetParamDefaultValueAsString(object? value, Type type, string typeRef)
...
+                => IsGenerationModeSpecified(typeSpec, JsonSourceGenerationMode.Metadata);
+
+            private static bool ShouldGenerateSerializationLogic(TypeGenerationSpec typeSpec)
+                => IsGenerationModeSpecified(typeSpec, JsonSourceGenerationMode.Serialization) && IsFastPathSupported(typeSpec);
+
+            private static bool IsGenerationModeSpecified(TypeGenerationSpec typeSpec, JsonSourceGenerationMode mode)
+                => typeSpec.GenerationMode == JsonSourceGenerationMode.Default || (mode & typeSpec.GenerationMode) != 0;
+
+            public static bool TryFilterSerializableProps(
```
Comment (layomia) [inline src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:1407]: We discussed before that this should be precomputed by the parser...

File: src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/CompilationHelper.cs [hunk 27]
```diff
@@ -12,15 +12,37 @@
...
+                    AllGeneratedTypes,
+                    spec => spec.TypeRef.FullyQualifiedName == fullyQualifiedName);
+    }
+
     public class CompilationHelper
     {
         private static readonly CSharpParseOptions s_parseOptions =
             new CSharpParseOptions(kind: SourceCodeKind.Regular, documentationMode: DocumentationMode.Parse);
 
+#if ROSLYN4_0_OR_GREATER
```
Comment (layomia) [inline src/libraries/System.Text.Json/tests/System.Text.Json.SourceGeneration.Unit.Tests/CompilationHelper.cs:42]: Does this PR provide an opportunity to switch to some helpers in `RoslynTestUtils`? https://github.com/dotnet/runtime/blob/main/src/libraries/Common/tests/SourceGenerators/RoslynTestUtils.cs

File: src/libraries/System.Text.Json/gen/Model/TypeRef.cs (hunk 23 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/Model/TypeRef.cs:26]: It would be inefficient because comparing the fully qualified name should suffice for types.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs (hunk 24 above)
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Parser.cs:392]: Will fix in a follow-up PR.

File: src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs [hunk 28]
```diff
@@ -111,18 +105,26 @@ public Emitter(in JsonSourceGenerationContext sourceGenerationContext, SourceGen
...
                     _generateGetConverterMethodForTypes = false;
                     _generateGetConverterMethodForProperties = false;
 
-                    foreach (TypeGenerationSpec typeGenerationSpec in _currentContext.RootSerializableTypes)
+                    Debug.Assert(_typeIndex.Count == 0);
+
+                    foreach (TypeGenerationSpec spec in _currentContext.GeneratedTypes)
                     {
-                        GenerateTypeInfo(typeGenerationSpec);
+                        _typeIndex.Add(spec.TypeRef, spec);
```
[the remark this answers is not in the document]
Comment (eiriktsarpalis) [reply src/libraries/System.Text.Json/gen/JsonSourceGenerator.Emitter.cs:124]: Will address in a follow-up.

**Comment by chsienki** [comment]
> > Have you validated this has the desired impact on recomputation in VS?
> 
> Yes, both in unit testing and debugging VS itself. Curiously enough though it appears to be re-triggering periodically every ~20 seconds or so regardless of whether it is needed or not. @chsienki @sharwell is this expected behavior?

When you say re-triggering, do you mean the whole generator is running, or just the inputs, and the generator finishes early because everything is cached?

**Comment by eiriktsarpalis** [comment]
> When you say re-triggering, do you mean the whole generator is running

The whole thing E2E.
